### PR TITLE
Replace stb_vorbis with libvorbis

### LINKED ISSUE #26964: Godot Editor crash when re-importing .ogg song that is currently playing
**Godot version:**
Godot 3.1 RC2

**OS/device including version:**
Win 7 Pro 64-bit, Win 8.1 Pro 64-bit

**Issue description:**
When an ogg file is playing, if you re-import it while it's playing, it freezes Godot and crashes it upon scene saving.

**Steps to reproduce:**
Open the project in Godot 3.1 RC2, High-light the AudioStreamPlayer, check the 'Playing' button in the Inspector, Highlight the .ogg file in the Filesystem window, go to the import window and change a setting and re-import the file.  This will cause only the 'Playing' button in the inspector to become unresponsive.  Further, if you try to save the scene, Godot will crash.

**Minimal reproduction project:**
[OggMusicTest.zip](https://github.com/godotengine/godot/files/2954141/OggMusicTest.zip)

### PR DESCRIPTION
**Pull request by ellenhp**
Fixes a variety of crashing bugs in stb_vorbis, and I think it also fixes #26964 but I will have to do a bit more testing to confirm that.

This introduces a size penalty of approximately 5% over the old internal representation of ogg files, but I think getting rid of stb_vorbis is worth it.

Important: If this gets merged and we ever want to revert it or switch to a new library or something, it's critical that we keep the change I made to editor_file_system.cpp otherwise whatever `.ogg` importer we replace this one with won't be run, and the engine will just crash instead.

### CODE REVIEW

File: modules/vorbis/audio_stream_ogg_vorbis.h [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+	// Seeks to the page such that the previous page has a granule position less than or equal to this value,
+	// and the current page has a granule position greater than this value.
+	// Returns true on success, false on failure.
+	bool seek_page(int64_t p_granule_pos);
+
+	OggPacketSequencePlayback() {}
+	virtual ~OggPacketSequencePlayback() {}
+};
+
+class AudioStreamOggVorbis;
```
Comment (aaronfranke) [inline modules/vorbis/audio_stream_ogg_vorbis.h:128]: The current classes have `OGG`. I'm not sure which is desired (is it even an acronym...?), but this needs discussion.

**Review by aaronfranke — COMMENTED** [review]
You need to update the documentation for the CI checks to pass.

File: modules/vorbis/resource_importer_ogg_vorbis.h [hunk 2]
```diff
@@ -28,31 +28,35 @@
...
 	virtual bool get_option_visibility(const String &p_option, const Map<StringName, Variant> &p_options) const override;
 
 	virtual Error import(const String &p_source_file, const String &p_save_path, const Map<StringName, Variant> &p_options, List<String> *r_platform_variants, List<String> *r_gen_files = nullptr, Variant *r_metadata = nullptr) override;
 
-	ResourceImporterOGGVorbis();
+	ResourceImporterOggVorbis();
 };
 
-#endif // RESOURCEIMPORTEROGGVORBIS_H
+#endif // RESOURCEIMPORTEROPUS_H
```
Comment (aaronfranke) [inline modules/vorbis/resource_importer_ogg_vorbis.h:62]: This isn't Opus.

Comment (ellenhp) [reply modules/vorbis/resource_importer_ogg_vorbis.h:62]: Yeah, sorry. This came out of some work I was doing to support opus and some stuff didn't get changed. It's still a draft, not really ready for a deep review yet.

**Comment by reduz** [comment]
This is looking extremely good so far!

**Comment by ellenhp** [comment]
I switched the names around, moved OGGPacketSequence to its own file in the ogg module, switched to using cache-friendlier data structures for the ogg packet data, and generally cleaned things up a bit. What this PR really needs right now is testing, not necessarily an in-depth review. Happy to respond to comments but what I'm most concerned about right now is it not actually working with some ogg files due to something I missed in the spec. It works with anything created with ffmpeg from my experience.

I'll put this in the PR description too but storing the ogg bitstream as a list of packets in a variant array does result in a size penalty compared to storing it unprocessed. The penalty is fairly severe before compression (25%+), but seems to be about 2-4% after re-compressing the saved `oggvorbisstr` file in a zip. I assume that during exports there will be compression applied so I'm hopeful this isn't a big issue?

**Comment by ellenhp** [comment]
I think (hope?) I've addressed all the CI issues. This should be ready to review now.

**Comment by Calinou** [comment]
> I assume that during exports there will be compression applied so I'm hopeful this isn't a big issue

PCK files don't use compression to speed up random file access, but ZIP archives exported by Godot do. Nonetheless, the vast majority of users use PCK files.

**Comment by ellenhp** [comment]
> Nonetheless, the vast majority of users use PCK files.

Let's wait to merge this until I figure out how to reduce the size penalty. I think I have some ideas.

**Comment by ellenhp** [comment]
The size penalty for the new imported type (oggvorbisstr) now seems to be around 5% over the old imported type (oggstr), which is pretty good.

I'm discarding some information to accomplish this then reconstructing it:

* b_o_s (beginning of stream): I can reconstruct this perfectly because each OGGPacketSequence contains only one logical stream, so this is 1 for the first page, 0 otherwise.
* e_o_s (end of stream): Similarly to b_o_s, I'm reconstructing this by looking at the position in the stream of any given page.
* granule_pos: I store this unmodified. I need it for seeking so I can't discard it.
* packetno: I reconstruct this by starting at zero for the first packet popped out of the OGGPacketSequence, then incrementing it for each packet. I reset it to zero whenever a seek is performed. This works, but it's not compliant with the standard I'm sure. I went through the libvorbis source code and it's only ever used for detecting holes/missing packets in the stream as far as I can tell, so I think what I'm doing is fine. The value isn't important, all that matters is that it increments by 1 for each contiguous packet.

There's one additional caveat to this PR:
* The importer and stream code currently only support ogg vorbis files with a single logical stream, so "chained" streams will only partially import. All logical streams after the first will be discarded. I don't know how common these files are in practice, but I suspect they're very uncommon.

**Comment by reduz** [comment]
Looks really good! Please squash commits so I can review and eventually merge.

**Review by reduz — APPROVED** [review]

**Comment by reduz** [comment]
This looks amazing, awesome job!

**Pull request merged**